mantidproject/mantid
# Create algorithm to load crystal structure from CIF

This addresses #11268.

**Changes**
An algorithm has been added that takes a `Workspace` and a CIF file as input and tries to assign the contained crystal structure to the Workspace's `Sample`-object. Additionally it can also load a UB-matrix from the file and apply it using `SetUB`. It uses [PyCifRW](https://pypi.python.org/pypi/PyCifRW/4.1) for parsing the CIF file, so this needs to be installed on the machine (normally through pip). If it's not, the algorithm is not registered into the factory and also the unit tests are skipped, the same way this is done for other algorithms with special dependencies as well.

There is much, much more information in CIF-files that could be used later on, but for now it only loads the parts that are absolutely necessary to create a `CrystalStructure`-object.

**Testing information**
1. Install `PyCifRW` if not already installed.
2. Get some CIF-files for testing. An excellent source if you don't have any available is the [Crystallography Open Database](http://www.crystallography.net), where many structures are placed in the public domain. You could go to "Browse", for example [the year 2014](http://www.crystallography.net/result.php?year=2014) and download a few CIFs from there (just click the "CIF" link in the table). Because the alternative space group settings are not available yet, select only files without any `:R` or `:2`, these won't work yet (so might some of the orthorhombic space groups).
3. Create a workspace, for example using `CreateSingleValuedWorkspace`. `TableWorkspace` does not work, because there is no `Sample` in these.
4. Set log level to "information".
5. Use `LoadCIF` to load the files you obtained in 2. Observe that the crystal structure it reports in the log corresponds to the one in the file (that means, unit cell, space group and atoms must match).
6. Manually test the usage example in the docs. This is not executed automatically as `PyCifRW` is not an official dependency of Mantid and thus not installed on the build servers.

Unfortunately I have not discovered a file with UB-information in that database yet, but this is also covered by the unit test, so I would not worry about this too much. Currently the UB-matrix will most likely come from another source anyway, this was just added for convenience.
